### ISSUE TITLE
Filter with startswith/endswith

### ISSUE THREAD

**Issue by ja-ne** [SAP/cloud-sdk-js#1055]
Hello,
using the SAP Cloud SDK (v. 1.38.0) to apply the filter function startswith (or endswith) the code may look like this:
```
import { filterFunctions } from '@sap-cloud-sdk/core';

const destination = getDestinationFromSomewhere();
SomeEntity.requestBuilder()
.getAll()
.filter(filterFunctions.startswith(SomeEntity.SOME_PROPERTY, 'm').equals(true))
.execute(destination);
```
This is the resulting URL (shortended for simplicity):
`https://<host>:<port>sap/opu/odata/sap/<service-name>/SomeEntity?$filter=(endswith(SomeProperty, 'm') eq true)`

Some properties may require the filter to be used as follows (without "eq true"):
`https://<host>:<port>sap/opu/odata/sap/<service-name>/SomeEntity?$filter=endswith(SomeProperty, 'm')`

Is it possible to build this URL using the SAP Cloud SDK?

Thank you in advance.

**Comment by marikaner** [comment]
Hey @ja-ne,

currently this is not possible. We have discussed the possibility to allow omitting `.equals(true)` in the application code, but not in the resulting URL as this should be logically the same. Do you have a service where `eq true` causes issues?
And if so, how would you express the opposite. Does `eq false` also fail, do you have to change it to `not()`?

**Comment by ja-ne** [comment]
Hi @marikaner,

the ODATA service API_PRODUCT_SRV throws an error when filtering for units (for example BaseUnit, WeightUnit, VolumeUnit) and attaching `eq true`. But I'm sure I'll find a solution.

I have not thought about expressing the opposite.

Thanks for your support.

**Comment by FrankEssenberger** [comment]
Hello @ja-ne,

that the `API_PRODUCT_SRV` throws an error is unfortunate. We had problems like this already in the past, that SAP services do not implement the OData spec correctly. If you look at the examples here: 

https://www.odata.org/documentation/odata-version-2-0/uri-conventions/

you see something like `endswith(CompanyName, 'Futterkiste') eq true`. So we can not change the behavior in general, but if we get more issues like this we will consider making the resulting `URL` adjustable for services which are important but do not honor the OData specifications.

Also feel free to create a ticket for the colleagues of the `API_PRODUCT_SRV` to handle the request correctly. I would close the issue? and hope you find a solution with the inversion.

**Comment by ja-ne** [comment]
Hello @FrankEssenberger,

many thanks for your ideas.

I had already created a ticket for the `API_RPODUCT_SRV`. After reading your answer now I think that for reasons I do not know the service expects startswith/endswith according to ODATA [v3](https://www.odata.org/documentation/odata-version-3-0/url-conventions/) or [v4](http://docs.oasis-open.org/odata/odata/v4.0/os/part2-url-conventions/odata-v4.0-os-part2-url-conventions.html#_Toc372793816).

Therefore I tried the `filterFunctionsV4` instead of `filterFunctionsV2` as a workaround but the result was the same. This solution may be quicker than checking/waiting for the service. Should `filterFunctionsV4.startswith` and, respectively, `filterFunctionsV4.endswith` build a filter like `$filter=endswith(CompanyName,'Futterkiste')` or am I missing something?

**Comment by FrankEssenberger** [comment]
Hello @ja-ne,

no I think you missed nothing and I think there is a bug also in our code. The `startsWith` and `endsWith` should not be in the [shared list](https://github.com/SAP/cloud-sdk-js/blob/fac2db1d1aa1693e5550641de34d256c8c72b182/packages/core/src/odata-common/filter/filter-functions.ts#L265-L274) but in the v2 or v4 implementaiton.

@marikaner @jjtang1985 do I see the issue correct because should be an easy fix to add the right implementation witout the `eq true` for v4.

**Comment by jjtang1985** [comment]
I tested with the `TripPin` service, and it seems both requests below work with the same response.
```
https://services.odata.org/TripPinRESTierService/People('russellwhyte')?$filter=startswith(UserName, 'russellwhyt') eq true
https://services.odata.org/TripPinRESTierService/People('russellwhyte')?$filter=startswith(UserName, 'russellwhyt')
```

I think we shared the same implementation between v2 and v4 based on this.

**Comment by ja-ne** [comment]
Hello @jjtang1985,

thanks for testing. There are a lot of cases where both requests work. But there seems to be a difference when working with units in the API_RPODUCT_SRV service:
`https://<host>:<port>/sap/opu/odata/sap/API_PRODUCT_SRV/A_Product?$filter=startswith(BaseUnit, 'G')` works while
`https://<host>:<port>/sap/opu/odata/sap/API_PRODUCT_SRV/A_Product?$filter=startswith(BaseUnit, 'G') eq true` fails with an error message.

**Comment by jjtang1985** [comment]
> Hey @ja-ne,
> 
> currently this is not possible. We have discussed the possibility to allow omitting `.equals(true)` in the application code, but not in the resulting URL as this should be logically the same. Do you have a service where `eq true` causes issues?
> And if so, how would you express the opposite. Does `eq false` also fail, do you have to change it to `not()`?

Hi @ja-ne , 
have you tested this workaround? Does it work for you?

By the way, the service `API_PRODUCT_SRV` is an OData V2 service.

**Comment by FrankEssenberger** [comment]
@jjtang1985 the use of the v4 filter function would be a work around in this case since the service does not accept the request as it should be. Besides that I think we have a bug in the filter function implementations for v4 since they are not to the specification. You are right, that for some services both versions are accepted but I think we can not rely on this.

**Comment by jjtang1985** [comment]
For v4 filter function, there are some examples "without eq true". Unfortunately, I did not find clear rule about boolean expression, which leads to some confusing that technically "with eq true" and "without eq true" should be the same but for the service owner, they might support one of them.

I also tested this [v2 reference service](https://services.odata.org/V2/Northwind/Northwind.svc/), where both cases, i.e., "with eq true" and "without eq true" work with the same response.

Unfortunately, however, the v2 documentation is even worse than v4, so I guess this is rather a OData common issue instead of v2/v4 specific.

I would suggest the following alternatives:
- improve the api about `BooleanFilterFunction`, so users can define the url patter (with/without eq true).
- support custom filter, which might be helpful for all filter related issues.

@ja-ne , it would be really great if you can share your test result with @marikaner 's workaround, so we understand whether this blocks your current work.

**Comment by ja-ne** [comment]
Sorry, I must have overlooked the workaround suggestion. I could not find the not() function exported by the SDK so I tested with the ODATA service itself.

If I am not mistaken the opposite of `endswith(SomeProperty, 'm') eq true` should be `not(endswith(SomeProperty,'m') eq false)`. Unfortunately this leads to the same error.

**Comment by jjtang1985** [comment]
@ja-ne ,

If `eq false` is not supported by the service, I would really doubt that whether filter function is implemented completely.

I tested this service via [API Hub](https://api.sap.com/) and the URLs below work as expected.
```
/sap/opu/odata/sap/API_PRODUCT_SRV/A_Product?$top=50&$filter=endswith(Product, '1') eq true
/sap/opu/odata/sap/API_PRODUCT_SRV/A_Product?$top=50&$filter=endswith(Product, '1') eq false
/sap/opu/odata/sap/API_PRODUCT_SRV/A_Product?$top=50&$filter=not(endswith(Product, '1') eq true)
```

Maybe you can share more information about the OData Service` API_PRODUCT_SRV`.
1. Is it a cloud system or On-Premise system?
2. Do you know which version it is?
3. Please share the original error from the OData Service if possible with one of the examples like `endswith(SomeProperty, 'm') eq false`. When testing it, please use e.g., browser or postman directly instead of the SDK to avoid potential bugs from the SDK.
4. Have you received any updates from the service owner, since the ticket has been created for 6 days.

Thank you very much.

Junjie

**Comment by ja-ne** [comment]
Hi @jjtang1985,
I do guess the filter is not implemented in the service in my case when filtering for units (for example BaseUnit, VolumeUnit, Weight Unit). But it could be some time before there is a fix so I thought a much faster solution would be using the filter function for ODATA v4.

I would not like to share more information openly.

**Comment by jjtang1985** [comment]
@ja-ne ,

I added a ticket to the backlog and hope we can handle it soon, since you cannot find a workaround and are blocked currently.
I'll let you know once there are further updates.

Best regards,
Junjie

**Comment by ja-ne** [comment]
@jjtang1985, thanks for your support.

**Comment by jjtang1985** [comment]
@artemkovalyov , the user mentioned [here](https://github.com/SAP/cloud-sdk-js/issues/1129) that this ticket should be treated as low priority.

**Comment by jjtang1985** [comment]
Hi @ja-ne ,
The filter functions with boolean return type can be used without `.equals(true)` like the example below:
```
SomeEntity.requestBuilder()
  .getAll()
  .filter(filterFunctions.startswith(SomeEntity.SOME_PROPERTY, 'm'))
  .execute(destination);
```
Also, the `not` function is implemented:
```
SomeEntity.requestBuilder()
  .getAll()
  .filter(not(filterFunctions.startswith(SomeEntity.SOME_PROPERTY, 'm')))
  .execute(destination);
```
These changes are available in the `canary` version, feel free to try it.

**Comment by jjtang1985** [comment]
I'll close this ticket, please re-open if you have questions.
Thank you.

**Comment by ja-ne** [comment]
Hi @jjtang1985.

thanks for the fix. I've tested and it works as expected.

**Comment by jjtang1985** [comment]
Hi all,
we released [1.42.0](https://github.com/SAP/cloud-sdk-js/releases), so you can switch to the new version for this fix.
